brave/brave-core
# NetworkService migration for Tor

brave-browser: a0b6c4d32833df4310609fd154405491c80e889f (75.0.3770.90)

fix https://github.com/brave/brave-browser/issues/4312

When NetworkService is enabled, we only have access to profile-wise
proxy config. So we discard proxy config per url request path to have a
global proxy config for tor profile and insert username/password when it
is about to resolve proxy in net::ProxyResolutionService.

Note
1. new flow works no matter NetworkService enabled or not.
2. ProxyConfigServiceTor has to be NET exported to avoid //net circular
  dependency.
3. ProxyConfigMonitor chromium overrides is for setting profile-wise
  proxy config without setting proxy_config::prefs::kProxy which will be
  interfered by other components(ex. DataReductionProxyService)

## Submitter Checklist:

- Verified that these changes build without errors on
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests && npm run test-security`) on

## Test Plan:
Launch Brave with/without `--enable-features=NetworkService` and test following scenario separately
1. Go to https://check.torproject.org/ in non tor window, make sure it doesn't connect to tor network
2. Go to https://check.torproject.org/ in tor window, remember the ip address
3. Click `New Tor Identity`, it will trigger a auto refresh and ip should be different than step 2

## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on.
